robosuite/models/objects/interactive_objects.py

```diff
@@ -0,0 +1,255 @@
+import robosuite.utils.env_utils as EU
+from copy import copy, deepcopy
+
+
+def satisfies(state, cond):
+    for k, v in cond.items():
+        if state[k] != v:
+            return False
+    return True
+
+
+class InteractiveObject(object):
+    """Base class for interactive object"""
+    def __init__(self, sim, body_id, cooldown_interval=-1):
+        self.sim = sim
+        self._cooldown_interval = cooldown_interval
+        self._last_event_step = 0
+        self._body_id = body_id
+
+    @property
+    def body_id(self):
+        return self._body_id
+
+    @property
+    def body_name(self):
+        return self.sim.model.body_id2name[self._body_id]
+
+    @property
+    def body_geom_ids(self):
+        return EU.bodyid2geomids(self.sim, self._body_id)
+
+    def reset(self):
+        self._last_event_step = 0
+
+    def _on_contact(self, contacting_body_ids):
```

Comment:
having trouble with understanding what exactly the difference is between InteractiveObject and StatefulObject. For example, what is the difference between the on_contact here vs. enter and leaving state? Are you factorizing functionality into doing things when enter / leave contact (i.e. setting state) vs. enter / leave state (i.e. state change)?

Reply:
states and events that cause states to change are different things. StatefulObject maintains object state and calls function when the state changes, but doesn't decide what event would change the state. InteractiveObject defines basic callback functions, and the sub-classes would determine how each event would change the state.